### PR TITLE
MINIFI-324 - More flexible uid generation

### PR DESCRIPTION
**Pull request by brosander**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [X] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] - N/A - If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [X] - N/A - If applicable, have you updated the LICENSE file?
- [X] - N/A - If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -326,26 +326,44 @@ Additionally, users can utilize the MiNiFi Toolkit Converter (version 0.0.1 - sc
...
-      
+
+### UID generation
+
+MiNiFi need to generate many unique identifiers in the course of operations.  There are a few different uid implementations available that can be configured in minifi-uid.properties.
+
+Implementation for uid generation can be selected using the uid.implementation property values:
+1. time - use uuid_generate_time
+2. random - use uuid_generate_random
+3. uuid_default - use uuid_generate (will attempt to use uuid_generate_random and fall back to uuid_generate_time if no high quality randomness is available)
```
Comment (phrocker) [inline README.md]: This statement confused me, but I'm guessing that it's in the context of uid.implementation. 
The following is logged to debug if uid.implementation isn't defined ( let's say there is no uid properties file ) 
"Using uuid_generate_time implementation for uids."

Does it make sense to make clear that uuid_generate_time is used as the fallback if the file does not exist?

File: libminifi/src/provenance/Provenance.cpp [hunk 2]
```diff
@@ -34,9 +36,24 @@ namespace nifi {
 namespace minifi {
 namespace provenance {
 
+std::shared_ptr<utils::IdGenerator> ProvenanceEventRecord::id_generator_ = utils::IdGenerator::getIdGenerator();
+std::shared_ptr<logging::Logger> ProvenanceEventRecord::logger_ = logging::LoggerFactory<ProvenanceEventRecord>::getLogger();
+
 const char *ProvenanceEventRecord::ProvenanceEventTypeStr[REPLAY + 1] = { "CREATE", "RECEIVE", "FETCH", "SEND", "DOWNLOAD", "DROP", "EXPIRE", "FORK", "JOIN", "CLONE", "CONTENT_MODIFIED",
     "ATTRIBUTES_MODIFIED", "ROUTE", "ADDINFO", "REPLAY" };
 
+ProvenanceEventRecord::ProvenanceEventRecord(ProvenanceEventRecord::ProvenanceEventType event, std::__cxx11::string componentId, std::__cxx11::string componentType) {
```
Comment (brosander) [inline libminifi/src/provenance/Provenance.cpp]: Autocomplete fail, will fix

File: README.md (hunk 1 above)
Comment (brosander) [inline README.md]: Well, uuid_generate_time is a lot faster than the random version but it leaks information about which node and what time the uuid originate from, I thought that might be relevant info if there are privacy-conscious applications of MiNiFi.

https://linux.die.net/man/3/uuid_generate_time

File: libminifi/include/utils/Id.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+namespace minifi {
+namespace utils {
+
+class IdGenerator {
+ public:
+  void generate(uuid_t output);
+  void initialize(const std::shared_ptr<Properties> & properties);
+  
+  static std::shared_ptr<IdGenerator> getIdGenerator() {
+    static std::shared_ptr<IdGenerator> generator = std::shared_ptr<IdGenerator>(new IdGenerator());
```
Comment (brosander) [inline libminifi/include/utils/Id.h]: I wanted to make the constructor private that seems to require manual invocation of new

```
error: ‘org::apache::nifi::minifi::utils::IdGenerator::IdGenerator()’ is private
   IdGenerator();
   ^
In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
                 from /usr/include/c++/5/bits/allocator.h:46,
                 from /usr/include/c++/5/bits/stl_tree.h:64,
                 from /usr/include/c++/5/map:60,
                 from /home/bryan/Github/nifi-minifi-cpp/libminifi/include/core/logging/LoggerConfiguration.h:24,
                 from /home/bryan/Github/nifi-minifi-cpp/libminifi/src/core/logging/LoggerConfiguration.cpp:21:
/usr/include/c++/5/ext/new_allocator.h:120:4: error: within this context
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
```

File: README.md (hunk 1 above)
Comment (phrocker) [inline README.md]: Sorry, I think my question was poorly worded. You make clear what happens when the file exists, which is what I meant by "in the context of uid.implementation [existing]". 

My question was more about documenting what will be used if the uid configuration file and the options don't exist. It seemed from the log message, "Using uuid_generate_time implementation for uids." that we were simply relying on time and not random first and then time.

File: libminifi/src/utils/Id.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        deterministic_prefix_[i] = prefix_element;
+      }
+      incrementor_ = 0;
+    } else if ("time" == implementation_str) {
+      logger_->log_debug("Using uuid_generate_time implementation for uids.");
+    } else {
+      logger_->log_debug("Invalid value for uid.implementation (%s). Using uuid_generate_time implementation for uids.", implementation_str);
+    }
+  } else {
+    logger_->log_debug("Using uuid_generate_time implementation for uids.");
```
Comment (phrocker) [inline libminifi/src/utils/Id.cpp]: That documentation question was regarding this log statement and how it seems without the first conditional being true, we simply default to time irrespective of random entry.

File: README.md (hunk 1 above)
Comment (phrocker) [inline README.md]: I added a comment in Id.cpp to clarify the location in code I was referring to. Sorry for the confusion.

**Comment by phrocker** [comment]
Looks good will merge shortly.

**Pull request closed without merging**